### PR TITLE
Separate Authentication and Encryption Service

### PR DESCRIPTION
**Pull request by iead-cc**
### Description
In this pull request the Encrpytion Service and the Authentication Service are separated.

### Parent Issue
Issue #cyber-crypt-com/encryptonize-premium/issues/3

### Comments for the Reviewers
- Created new grpc service: authn.Encryptonize
- Created Authenticator Interface
- Moved Authstorage Middlewear into its propper place
- Still uses the old Token format not up to date with the Security Architecture

### Type(s) of Change (Split the PR if you check many)
- [ ] Added user feature
- [ ] Added technical feature
- [ ] Added tests
- [x] Refactor
- [ ] Bugfix
- [ ] Updated documentation
- [ ] Infrastructure changes (Terraform, GCP, K8s, other)
- [ ] CI/CD workflow changes

### Testing Checklist
- [ ] I have added/updated unit tests for functional changes.
- [x] I have added/updated integration tests for changes that affect dependent systems.
- [ ] I have added/updated end-to-end tests for feature changes.

### General Checklist
- [x] I have pulled in the latest changes from master.
- [x] I have run `make lint`.
- [x] I have successfully run `make docker-up; make tests`.
- [ ] I have updated relevant workflows and deployment tools.
- [ ] I have updated/added relevant documentation (readme, doc comments, etc).
- [x] I have added the license to new source code files.
- [x] I have spent some time looking over the full diff before creating this PR.

### Technical Debt
(If there is known technical debt added at the end of this change, please create a new (technical)
issue and link it here - otherwise delete this section)
Issue #xxx

### CODE REVIEW

File: encryption-service/app/app.go [hunk 1]
```diff
@@ -49,6 +50,7 @@ var GitTag string
 type App struct {
 	Config               *Config
 	MessageAuthenticator *crypt.MessageAuthenticator
+	Authenticator        authn.AuthenticatorInterface
```
Comment (bruno-cc) [inline encryption-service/app/app.go:53]: I really don't like the word `Authenticator` anymore. I think 
 `MessageAuthenticator` vs `Authenticator` is a really bad naming for different things.

File: encryption-service/app/app.go [hunk 2]
```diff
@@ -187,8 +189,16 @@ func CheckInsecure(config *Config) {
 
 // CreateAdminCommand creates a new admin users with random credentials
 // This function is intended to be used for cli operation
-func (app *App) CreateAdminCommand() {
+func (app *App) CreateAdminCommand() error {
 	ctx := context.Background()
+
+	auth, ok := app.Authenticator.(*authn.Authenticator)
+	if !ok {
+		err := errors.New("app.Authenticator is not of type authn.Authenticator")
```
Comment (bruno-cc) [inline encryption-service/app/app.go:197]: Can this even happen with the design of the `App` struct?

File: encryption-service/authn/authn_test.go [hunk 3]
```diff
@@ -11,40 +11,24 @@
...
-		MessageAuthenticator: messageAuthenticator,
-	}
-
-	expectedSerialized = "ChAAAAAAAABAAIAAAAAAAAACEgEE"
-	expectedMessage, _ = base64.RawURLEncoding.DecodeString(expectedSerialized)
-	expectedTag, _     = messageAuthenticator.Tag(crypt.TokenDomain, append(nonce, expectedMessage...))
-	expectedToken      = "ChAAAAAAAABAAIAAAAAAAAACEgEE.AAAAAAAAAAAAAAAAAAAAAg." + base64.RawURLEncoding.EncodeToString(expectedTag)
-)
-
+// copy
```
Comment (bruno-cc) [inline encryption-service/authn/authn_test.go:31]: I think that's confusing. Either add a TODO or remove I guess.

File: encryption-service/authn/token.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		return "", errors.New("Invalid userID UUID")
+	}
+
+	userScope := []UserScope{}
+	// scopes is a bitmap. This checks each bit individually
+	for i := ScopeType(1); i < ScopeEnd; i <<= 1 {
+		if !accessToken.UserScopes.HasScopes(i) {
+			continue
+		}
+		switch i {
```
Comment (bruno-cc) [inline encryption-service/authn/token.go:89]: I think we often have these mapping contraptions. Should we create a map for these?

File: encryption-service/authn/token.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+// value ensures unique user facing token even if the actual access token
+// would be equal. It also checks the validity of the access token as
+// this is last function every token has to go through before the token
+// are presented to an API. If this method only signs valid token we
+// can then assume that any signed token is valid.
+// The returned token has three parts. Each part is individually base64url encoded
+// the first part (data) is a serialized protobuf message containing
+// the user ID and a set of scopes. The structure of the assembled token is
+// <data>.<nonce>.HMAC(nonce||data)
+func (a *Authenticator) SerializeAccessToken(accessToken *AccessToken) (string, error) {
```
Comment (bruno-cc) [inline encryption-service/authn/token.go:69]: We already discussed that I am not a big fan of this Serialize function, but I we will rework it anyhow in the near future.

File: encryption-service/authstorage/authstorage_middleware.go [hunk 6]
```diff
@@ -39,11 +41,11 @@ func (app *App) AuthStorageUnaryServerInterceptor() grpc.UnaryServerInterceptor
 
 		// Don't start DB tranaction on health checks
```
Comment (bruno-cc) [inline encryption-service/authstorage/authstorage_middleware.go:42]:
```suggestion
		// Don't start DB transaction on health checks
```

File: encryption-service/authn/authn.go [hunk 7]
```diff
@@ -14,210 +14,96 @@
...
 // Authenticator represents a MessageAuthenticator used for signing and checking the access token
 type Authenticator struct {
 	MessageAuthenticator *crypt.MessageAuthenticator
+	UnimplementedEncryptonizeServer
 }
 
 type AuthenticatorInterface interface {
-	SerializeAccessToken(accessToken *AccessToken) (string, error)
-	ParseAccessToken(token string) (*AccessToken, error)
+	RegisterService(srv grpc.ServiceRegistrar)
```
Comment (bruno-cc) [inline encryption-service/authn/authn.go:37]: Not sure if that isn't  just wrapping for wrapping reasons.

File: encryption-service/authn/authn.go [hunk 8]
```diff
@@ -14,210 +14,96 @@
...
+		return nil, err
 	}
 
-	data, err := proto.Marshal(accessTokenClient)
-	if err != nil {
-		return "", err
+	// Don't authenticate health checks
+	// IMPORTANT! This check MUST stay at the top of this function
+	if methodName == health.HealthEndpointCheck ||
+		methodName == health.HealthEndpointWatch {
```
Comment (bruno-cc) [inline encryption-service/authn/authn.go:75]: I would keep that in the same line.

File: encryption-service/authn/authn_test.go (hunk 3 above)
Comment (iead-cc) [reply encryption-service/authn/authn_test.go]: fixed in 4478075

File: encryption-service/authn/authn.go (hunk 7 above)
Comment (iead-cc) [reply encryption-service/authn/authn.go:37]: A different Auth implementation may offer a different API.

File: encryption-service/authn/token.go (hunk 5 above)
Comment (iead-cc) [reply encryption-service/authn/token.go:69]: Tracked in #121 (more or less)

File: encryption-service/authn/token.go (hunk 4 above)
Comment (iead-cc) [reply encryption-service/authn/token.go:89]: we translate from protobuf twice, and to protobuf once

File: encryption-service/app/app.go (hunk 2 above)
Comment (iead-cc) [reply encryption-service/app/app.go:197]: If we use a different Implementation of the Authservice. But currently not. Also the error is slightly wrong...

File: encryption-service/authn/authn.go [hunk 9]
```diff
@@ -14,210 +14,95 @@
...
-			userScope = append(userScope, AccessTokenClient_USERMANAGEMENT)
-		default:
-			return "", errors.New("Invalid scopes")
-		}
-	}
-
-	accessTokenClient := &AccessTokenClient{
-		UserId:     accessToken.UserID.Bytes(),
-		UserScopes: userScope,
+// CheckAccessToken verifies the authententicity of a token and
```
Comment (bruno-cc) [inline encryption-service/authn/authn.go:58]:
```suggestion
// CheckAccessToken verifies the authenticity of a token and
```

File: encryption-service/authn/authn.go [hunk 10]
```diff
@@ -14,210 +14,95 @@
...
+// that the token contains the required token for the requested API
```
Comment (bruno-cc) [inline encryption-service/authn/authn.go:59]:
```suggestion
// that the token contains the required scope for the requested API
```

File: encryption-service/authn/token.go (hunk 4 above)
Comment (bruno-cc) [reply encryption-service/authn/token.go:89]: ok exaggerated.

File: encryption-service/app/app.go (hunk 2 above)
Comment (bruno-cc) [reply encryption-service/app/app.go:197]: nevermind I misread the cast, sorry!

Comment (iead-cc) [reply encryption-service/app/app.go:197]: I just looked at it and it cannot happen as the cast is checked earlier already

File: encryption-service/app/app.go (hunk 1 above)
Comment (iead-cc) [reply encryption-service/app/app.go]: renamed in 4616c42

File: encryption-service/authn/authn.go (hunk 10 above)
Comment (iead-cc) [reply encryption-service/authn/authn.go]: fixed in 59687df

File: encryption-service/authn/authn.go (hunk 9 above)
Comment (iead-cc) [reply encryption-service/authn/authn.go]: fixed in 59687df

File: encryption-service/authn/authn.go (hunk 8 above)
Comment (iead-cc) [reply encryption-service/authn/authn.go]: fixed in 59687df

**Review by bruno-cc — APPROVED** [review]
Looks good. Thank you!

**Pull request merged**